### PR TITLE
Add DTD version of sidebar

### PR DESCRIPTION
**Pull request by DanTup**
~~This is a work in progress to add a DTD implementation of the sidebar (while still maintaining a postMessage-compatible version for some transition period). There's still work to do here but I'm opening a Draft PR as a reference for what I'm working towards as I try to land smaller PRs:~~

This adds a DTD version of the sidebar alongside the current postMessage version. They implement a common interface so we can (mostly) have one implementation of the sidebar widgets while supporting both so that we can initially test the sidebar with a flag and then switch over once we're sure it's good.

Although it's mostly done, there are still a few things outstanding before this is mergable:

- [x] Merge latest DevTools master into this PR
- [x] Publish a new `package:dtd` with the latest changes and update the version here (we rely on some new constructors and type changes in package:dtd)
- [x] Ensure the DTD changes to support the Service stream have rolled into Flutter and that version of Flutter has rolled into DevTools (current is `3.5.0-293.0.dev` but we need something with [23de3198faaa624c68cce347574a531854699af5](https://github.com/dart-lang/sdk/commit/23de3198faaa624c68cce347574a531854699af5) which looks like `3.6.0-16.0.dev`)
- [x] Move the spec for the editor service to the SDK repo (in the doc @helin24 started)
- [x] Re-test both the simulated environment and in VS Code using `dart.customDevTools` (without dependency overrides etc., just using the pinned version of Flutter)

### CODE REVIEW

**Comment by DanTup** [comment]
@kenzieschmoll this is mostly ready now (I don't have any more changes planned, so it's at least ready for a round of feedback). I can load this version of the sidebar in VS Code and all of the functionality from the previous version works (and as a bonus of not using postMessage, we can open the sidebar outside of VS Code and it's still fully functional!).

It adds a new endpoint `/editorSidebar` for the DTD-based version (named more generically) which just wraps the same sidebar but providing a DTD connection instead of a `postMessage` API (they both implement the same interface). It also adds a new stager app for this DTD-based version (and some VS Code tasks/launch config to start a DTD connection for it).

Initially the Dart-Code changes will be landed behind a flag (`"dart.previewDtdSidebar": true,`) to allow some additional testing (and selection of the SDK version we want to gate it on).

There will be a fair bit of code we can clean up/merge once we get rid of the `postMessage` version, but we need to support both for now to allow us to switch over gradually (and be able to switch back in VS Code if we find issues).

There's a temporary markdown file here that will be moved to the SDK, but I'll wait for @helin24's change to land first because it'll go into the same file.

**Comment by DanTup** [comment]
(the bot failure is because we need a newer version of DTD that isn't published yet - there have still been some other changes ongoing in that, so it's easier to release once they're all done - that will need doing before we can merge this ofc)

File: packages/devtools_app/test/test_infra/scenes/standalone_ui/mock_editor_widget.dart
Comment (kenzieschmoll) [inline packages/devtools_app/test/test_infra/scenes/standalone_ui/mock_editor_widget.dart:86]: do these stream subs need to be cancelled? if so you can use the AutoDisposeMixin to add then as automatically disposed subscriptions

File: packages/devtools_app/test/test_infra/scenes/standalone_ui/mock_editor_widget.dart
Comment (kenzieschmoll) [inline packages/devtools_app/test/test_infra/scenes/standalone_ui/mock_editor_widget.dart:147]: super nit: add `const SizedBox(width: denseSpacing)` between these buttons

File: packages/devtools_app/test/test_infra/scenes/standalone_ui/mock_editor_widget.dart
Comment (kenzieschmoll) [inline packages/devtools_app/test/test_infra/scenes/standalone_ui/mock_editor_widget.dart:288]: How about a ListView.builder instead?

File: packages/devtools_app/test/test_infra/scenes/standalone_ui/mock_editor_widget.dart
Comment (kenzieschmoll) [inline packages/devtools_app/test/test_infra/scenes/standalone_ui/mock_editor_widget.dart:314]: same here

File: packages/devtools_app/test/test_infra/scenes/standalone_ui/utils.dart
Comment (kenzieschmoll) [inline packages/devtools_app/test/test_infra/scenes/standalone_ui/utils.dart:13]: nit: trailing comma

File: packages/devtools_app/test/test_infra/scenes/standalone_ui/utils.dart
Comment (kenzieschmoll) [inline packages/devtools_app/test/test_infra/scenes/standalone_ui/utils.dart:16]: does this channel need to be closed at some point by the caller of createLoggedWebSocketChannel? Otherwise we might be leaking resources

**Review by kenzieschmoll — APPROVED** [review]

**Comment by kenzieschmoll** [comment]
Whoops didn't mean to approve, just meant to comment. Undo LGTM.

File: packages/devtools_app/lib/src/service/editor/editor_client.dart
Comment (DanTup) [reply packages/devtools_app/lib/src/service/editor/editor_client.dart:91]: We can't remove the variable because it's part of the external interface (the sidebar will fetch the initial set of devices/debug sessions once this completes).

File: packages/devtools_app/lib/src/service/editor/editor_client.dart
Comment (DanTup) [reply packages/devtools_app/lib/src/service/editor/editor_client.dart:86]: `DisposerMixin` doesn't seem to have a `dispose()` method. I see one in `AutoDisposeControllerMixin` that calls `cancelStreamSubscriptions`, `cancelListeners`, `cancelFocusNodes` but it's not clear if I should use that here (it adds some additional stuff that doesn't seem relevant.

(I could just duplicate that `dispose()` method into my class, but it doesn't feel like the best thing)

File: packages/devtools_app/lib/src/service/editor/editor_client.dart
Comment (DanTup) [reply packages/devtools_app/lib/src/service/editor/editor_client.dart:129]: Unfortunately `EditorMethod.hotRestart.name` is not constant (`name` is an extension from the SDK). I did try doing this a few other ways but they all felt a little messy too (for example adding an explicit `name` to the enum values means every enum value has to have its name duplicated as a string in the constructor call). I can change if you have any preference between them though.

File: packages/devtools_app/test/test_infra/scenes/standalone_ui/editor_service/editor_server.dart
Comment (DanTup) [reply packages/devtools_app/test/test_infra/scenes/standalone_ui/editor_service/editor_server.dart:108]: It's possible - we might want to let the simulated editor set some workspace roots for extensions (I think currently we just mock them). Any objection to changing the `uri` from `String` to `Uri` in:

```dart
typedef DTDConnectionInfo = ({String? uri, String? secret});
```

Many of these places are already using `Uri`s and I feel like it might be better to change the other places to use `Uri`s than these to `String`s.

File: packages/devtools_app/test/test_infra/scenes/standalone_ui/editor_service/editor_server.dart
Comment (DanTup) [reply packages/devtools_app/test/test_infra/scenes/standalone_ui/editor_service/editor_server.dart:183]: I don't think so - we use `map` and `pipe` on those streams and it says those will pass the `done` events through (and closing `dtd` here calls `close` on the `Peer` that was given those streams.

File: packages/devtools_app/test/test_infra/scenes/standalone_ui/editor_service/post_message_fake_editor.dart
Comment (DanTup) [reply packages/devtools_app/test/test_infra/scenes/standalone_ui/editor_service/post_message_fake_editor.dart:92]: I think doing nothing is fine. It just means the Connect/Disconnect buttons in the simulated editor won't do anything for the postMessage version (and not throw), but hopefully this postMessage code isn't going to live long and this whole file will be deleted (I think we can do this much faster than when need to compatible in the other direction, since VS Code extensions tend to be updated fairly quickly).

File: packages/devtools_app/test/test_infra/scenes/standalone_ui/utils.dart
Comment (DanTup) [reply packages/devtools_app/test/test_infra/scenes/standalone_ui/utils.dart:16]: I believe this is the same as above. The stream gets passed to the `DartToolingDaemon` class and it will handle closing them (which is handled by the fake editor if you click the disconnect button, but in this test app we don't really have another way to close it - eg. you can't navigate away from the mock editor scene when you run this app).

File: packages/devtools_app/lib/src/service/editor/api_classes.dart
Comment (DanTup) [reply packages/devtools_app/lib/src/service/editor/api_classes.dart]: Added backticks in 77f796e9c (can't use brackets as there's nothing in-scope with that name here).

File: packages/devtools_app/lib/src/service/editor/editor_client.dart
Comment (DanTup) [reply packages/devtools_app/lib/src/service/editor/editor_client.dart]: Done in 77f796e9c

File: packages/devtools_app/lib/src/service/editor/editor_client.dart
Comment (DanTup) [reply packages/devtools_app/lib/src/service/editor/editor_client.dart]: Done in 77f796e9c

File: packages/devtools_app/lib/src/service/editor/editor_client.dart
Comment (DanTup) [reply packages/devtools_app/lib/src/service/editor/editor_client.dart]: I thought we don't, since we really just care when it changed and then we can read the boolean flags (perhaps once the postMessage version is done we could use the capabilities directly from the event and toggle state in the widget). However, you made me look at this again and I saw that we would treat the unregistration of `editor.GetDevices` as if it was a registration and then try to call it. Hopefully that never happens (the editor shouldn't be disconnecting from DTD while the sidebar is visible!), but I've updated this in 83fd3e7df to be more structured (the stream is one of two subclasses for registration/unregistration with additional fields).

File: packages/devtools_app/lib/src/service/editor/editor_client.dart
Comment (DanTup) [reply packages/devtools_app/lib/src/service/editor/editor_client.dart]: Done in 958878af8

File: packages/devtools_app/lib/src/standalone_ui/standalone_screen.dart
Comment (DanTup) [reply packages/devtools_app/lib/src/standalone_ui/standalone_screen.dart:32]: Done in b65f2bb0e

File: packages/devtools_app/lib/src/standalone_ui/vs_code/flutter_panel.dart
Comment (DanTup) [reply packages/devtools_app/lib/src/standalone_ui/vs_code/flutter_panel.dart]: Done in b65f2bb0e

File: packages/devtools_app/lib/src/standalone_ui/vs_code/flutter_panel.dart
Comment (DanTup) [reply packages/devtools_app/lib/src/standalone_ui/vs_code/flutter_panel.dart]: Correct. I presume you don't need to set up anything in analytics to handle the new value? I've used `editorSidebar` (and left the original one as `vsCodeFlutterSidebar`... I did initially rename it to `legacyVsCodeFlutterSidebar` but then realised it would fragment new stats from the previous ones so reverted that).

File: packages/devtools_app/test/test_infra/scenes/standalone_ui/editor_service/editor_server.dart
Comment (DanTup) [reply packages/devtools_app/test/test_infra/scenes/standalone_ui/editor_service/editor_server.dart]: Done in b65f2bb0e

File: packages/devtools_app/test/test_infra/scenes/standalone_ui/editor_service/editor_server.dart
Comment (DanTup) [reply packages/devtools_app/test/test_infra/scenes/standalone_ui/editor_service/editor_server.dart]: Done in b65f2bb0e

File: packages/devtools_app/test/test_infra/scenes/standalone_ui/editor_service/fake_editor.dart
Comment (DanTup) [reply packages/devtools_app/test/test_infra/scenes/standalone_ui/editor_service/fake_editor.dart]: Done in b65f2bb0e

File: packages/devtools_app/test/test_infra/scenes/standalone_ui/editor_sidebar.dart
Comment (DanTup) [reply packages/devtools_app/test/test_infra/scenes/standalone_ui/editor_sidebar.dart]: Turns out they're not anymore. I think the sidebar/standalone pages used to be behind an experiment and this was initially a copy/paste of the postMessage version. I've removed the flag from the comments in both.

File: packages/devtools_app/test/test_infra/scenes/standalone_ui/editor_sidebar.dart
Comment (DanTup) [reply packages/devtools_app/test/test_infra/scenes/standalone_ui/editor_sidebar.dart]: Done in b65f2bb0e

File: packages/devtools_app/test/test_infra/scenes/standalone_ui/mock_editor_widget.dart
Comment (DanTup) [reply packages/devtools_app/test/test_infra/scenes/standalone_ui/mock_editor_widget.dart]: Ah nice, that's much better. Added in d50c807fe

File: packages/devtools_app/test/test_infra/scenes/standalone_ui/mock_editor_widget.dart
Comment (DanTup) [reply packages/devtools_app/test/test_infra/scenes/standalone_ui/mock_editor_widget.dart]: Done in d50c807fe

File: packages/devtools_app/test/test_infra/scenes/standalone_ui/mock_editor_widget.dart
Comment (DanTup) [reply packages/devtools_app/test/test_infra/scenes/standalone_ui/mock_editor_widget.dart:147]: Done in e07fca26a (here and below)

File: packages/devtools_app/test/test_infra/scenes/standalone_ui/mock_editor_widget.dart
Comment (DanTup) [reply packages/devtools_app/test/test_infra/scenes/standalone_ui/mock_editor_widget.dart]: Done in e07fca26a

File: packages/devtools_app/test/test_infra/scenes/standalone_ui/utils.dart
Comment (DanTup) [reply packages/devtools_app/test/test_infra/scenes/standalone_ui/utils.dart]: Still looked naff with a trailing comma, so I pulled the return type out into a typedef to avoid the wrapping in d50c807fe

File: packages/devtools_app/test/test_infra/scenes/standalone_ui/editor_sidebar.dart
Comment (DanTup) [reply packages/devtools_app/test/test_infra/scenes/standalone_ui/editor_sidebar.dart]: It does not currently - but it's probably something we'll want to support. I'll add a TODO because I'm not sure exactly what we want here (it would be nice if you could set the workspace roots from buttons in the simulated editor) and this PR is already quite large.

(TODO added in 60707d74c)

File: packages/devtools_app/test/test_infra/scenes/standalone_ui/editor_service/editor_server.dart
Comment (DanTup) [reply packages/devtools_app/test/test_infra/scenes/standalone_ui/editor_service/editor_server.dart]: I've renamed these so that "FakeEditor" and "EditorServer" (and similar names) are now "SimulatedEditor" and "SimulatedEditorBase" in 60707d74c.

These classes are split up more than they'll ultimately need to be to support having two implementations (and the idea that EditorServer might be useful outside of DevTools). We can merge/flatten some of these classes when we remove the postMessage versions that I think will probably help clarify things much more too (there's a TODO about this).

**Comment by DanTup** [comment]
@kenzieschmoll thanks! I've been through and addressed (or responded to) all of your points now (sorry for the spam).

I've also edited in a checklist to the top comment here with some final things that will need doing before this can merge (the bots will continue to be red until some more of those are done).

File: packages/devtools_app/lib/src/service/editor/editor_client.dart
Comment (kenzieschmoll) [reply packages/devtools_app/lib/src/service/editor/editor_client.dart:90]: if EditorClient extends DisposeableController, then you can mixin AutoDisposeControllerMixin here.

File: packages/devtools_app/lib/src/standalone_ui/vs_code/flutter_panel.dart
Comment (kenzieschmoll) [reply packages/devtools_app/lib/src/standalone_ui/vs_code/flutter_panel.dart]: Nope we will just start getting events with the new string. And yes, you are correct we should leave the legacy name as it was so we don't fragment the data

File: packages/devtools_app/test/test_infra/scenes/standalone_ui/editor_service/editor_server.dart
Comment (kenzieschmoll) [reply packages/devtools_app/test/test_infra/scenes/standalone_ui/editor_service/editor_server.dart]: I believe this would be a breaking change to devtools_shared, so would prefer that we leave it as is for now and we can add a TODO to consider doing this later.

File: packages/devtools_app/test/test_infra/scenes/standalone_ui/vs_code.dart
Comment (kenzieschmoll) [inline packages/devtools_app/test/test_infra/scenes/standalone_ui/vs_code.dart:24]: remove leading space

File: packages/devtools_app_shared/lib/src/utils/list_queue.dart
Comment (kenzieschmoll) [inline packages/devtools_app_shared/lib/src/utils/list_queue.dart:21]: this should have tests. To not block this PR, I recommend putting this class under devtools_app/ and then adding a matching test file devtools_app/test/shared/ with a TODO and a tracking issue. Once this class is well tested we can consider adding to the public package devtools_app_shared/ for shared use.

File: packages/devtools_app/test/test_infra/scenes/standalone_ui/editor_service/editor_server.dart
Comment (DanTup) [reply packages/devtools_app/test/test_infra/scenes/standalone_ui/editor_service/editor_server.dart]: Ah yes, ofc.

I realised when I went back to this that getting a secret to here is going to be complicated because for now we're just spawning the DTD instance as a VS Code task - we don't have a way to get the secret over here. I'll need to think a bit more about this so will leave for another PR.

File: packages/devtools_app/lib/src/service/editor/editor_client.dart
Comment (DanTup) [reply packages/devtools_app/lib/src/service/editor/editor_client.dart]: Ah, done in 1bd6ef80d

File: packages/devtools_app/lib/src/service/editor/editor_client.dart
Comment (DanTup) [reply packages/devtools_app/lib/src/service/editor/editor_client.dart]: Done in 98d75b839

File: packages/devtools_app/lib/src/shared/analytics/constants/_editor_sidebar_constants.dart
Comment (DanTup) [reply packages/devtools_app/lib/src/shared/analytics/constants/_editor_sidebar_constants.dart]: I don't think so - the original code had an entry here and the static wrapping it, but seems like this was the only one - so I've removed both of the enum values that are just the sidebar IDs and put the literals into the statics instead (done in 98d75b839)

File: packages/devtools_app/test/test_infra/scenes/standalone_ui/editor_sidebar.dart
Comment (DanTup) [reply packages/devtools_app/test/test_infra/scenes/standalone_ui/editor_sidebar.dart]: Done in 98d75b839

File: packages/devtools_app_shared/lib/src/utils/list_queue.dart
Comment (DanTup) [reply packages/devtools_app_shared/lib/src/utils/list_queue.dart]: Done in 98d75b839

File: packages/devtools_app/test/test_infra/scenes/standalone_ui/editor_sidebar.dart
Comment (DanTup) [reply packages/devtools_app/test/test_infra/scenes/standalone_ui/editor_sidebar.dart:80]: Multiline TODOs are indented by a space because that's how the analyzer knows the next line is part of the TODO and not an unrelated comment. If we remove this, the TODO diagnostics will be truncated. It might not matter too much here if it's unlikely anyone is using the TODO diagnostics though, so I'll do whichever you prefer!

**Comment by DanTup** [comment]
@kenzieschmoll thanks! I've addressed/responded to your latest comments. I also cleared your review status because I think it was your accidental approve rather than a new one.

I'll try to progress the other things in the checklist above that this PR depends on.

**Review by kenzieschmoll — APPROVED** [review]
looks good!

**Comment by DanTup** [comment]
I've merged latest master into this and the bots have gone green. I've removed the temp spec file here and opened https://dart-review.googlesource.com/c/sdk/+/378141 to put it in the SDK with the other common service interfaces. I also made a minor tweak to support `null` `deviceId`s for `selectDevice` (VS Code already supported it and we had it in the event, but not the request).

I've also tested this latest version again with the Dart-Code branch that uses the DTD sidebar and everything is working:

image

@kenzieschmoll any final comments before I merge this? I will merge the Dart-Code work after tomorrows release so it will become available behind a flag in the pre-release extension tomorrow so it'll be easy for others to test (and we can figure out when to remove the flag and just gate on SDK version when we're sure it's all good).

**Comment by kenzieschmoll** [comment]
Let's hold off until the 2.38.0 release of DevTools is complete (working on that today). Will ping this PR or just merge it once the release is out. Thanks!

**Comment by kenzieschmoll** [comment]
The 2.38.0 DevTools release into the SDK is complete. Good to merge this now.

**Pull request merged**
